src/peer/socket.rs

```diff
@@ -99,103 +91,107 @@ async fn retainer<S: Signer>(
-                break 'receive;
+                if let Err(_err) = down_tx.send(message).await {
+                    log::error!("failed to push received message");
+                }
             }
+        });
 
+        'receive: loop {
             tokio::select! {
-                Some(message) = up_rx.recv() => {
-                    log::trace!("sending message to relay {:?}", &u.domain());
-                    if let Err(err) = write.send(message).await {
-                        // probably connection closed as well, we need to renew!
-                        log::error!("disconnected: error while sending message: {}", err);
-                        break 'receive;
+                read_err = closed.recv() => {
+                    if let Some(err) = read_err {
+                        log::error!("read error: {}", err);
                     }
+                    log::debug!("read routine exited, reconnecting");
+                    // force reconnecting
+                    break 'receive;
                 },
-                message = read.recv() => {
-                    let message = match message {
-                        None=> {
-                            log::error!("disconnected: read stream ended")  ;
-                            break 'receive;
-                        },
-                        Some(message) => message,
-                    };
+                message = timeout(PING_INTERVAL, up_rx.recv()) => {
+                    // first check if we have timed out
+                    let now = timestamp();
+                    if now - ts.load(Ordering::Relaxed) > READ_TIMEOUT {
+                        // we have timed out! we need to reconnect then
+                        log::info!("connection timeout! retrying");
+                        break 'receive
+                    }
 
-                    // we take a note with when a message was received
-                    last = Instant::now();
-                    log::trace!("received a message from relay {:?}", &u.domain());
                     let message = match message {
-                        Ok(Message::Pong(_)) => {
-                            log::trace!("received a pong");
-                            continue 'receive;
-                        }
-                        Ok(message) => message,
-                        Err(err) => {
-                            // todo: those errors probably mean we need to re-connect
-                            // we will see what to do later.
-                            log::error!("disconnected: error while receiving message: {}", err);
-                            break 'receive;
+                        Ok(Some(message)) => message,
+                        Ok(None) => {
+                            //weird why would we receive a non message
+                            log::error!("socket closed!");
+                            handler.abort();
+                            return
```

Comment:
This is a way to terminate the connection. This can only be none if the sender part of a mpsc channel is dropped, this can only happen if the `Socket` structure that we have is dropped. If that happen it's totally find to drop the entire connection, and return.